lib/datadog/core/telemetry/collector.rb

```diff
@@ -0,0 +1,207 @@
+# typed: true
+
+require 'etc'
+
+require 'datadog/core/diagnostics/environment_logger'
+require 'datadog/core/environment/ext'
+require 'datadog/core/environment/platform'
+require 'datadog/core/telemetry/v1/application'
+require 'datadog/core/telemetry/v1/appsec'
+require 'datadog/core/telemetry/v1/configuration'
+require 'datadog/core/telemetry/v1/dependency'
+require 'datadog/core/telemetry/v1/host'
+require 'datadog/core/telemetry/v1/integration'
+require 'datadog/core/telemetry/v1/product'
+require 'datadog/core/telemetry/v1/profiler'
+
+module Datadog
+  module Core
+    module Telemetry
+      # Module defining methods for collecting metadata for telemetry
+      # rubocop:disable Metrics/ModuleLength
+      module Collector
+        include Datadog::Core::Configuration
+
+        # Forms a telemetry application object
+        def application
+          Telemetry::V1::Application
+            .new(
+              env: env,
+              language_name: Datadog::Core::Environment::Ext::LANG,
+              language_version: Datadog::Core::Environment::Ext::LANG_VERSION,
+              products: products,
+              runtime_name: Datadog::Core::Environment::Ext::RUBY_ENGINE,
+              runtime_version: Datadog::Core::Environment::Ext::ENGINE_VERSION,
+              service_name: service_name,
+              service_version: service_version,
+              tracer_version: tracer_version
+            )
+        end
+
+        # Forms a telemetry configurations object
+        def configurations
+          configuration_variables
+        end
+
+        # Forms a telemetry app-started dependencies object
+        def dependencies
+          Gem::Specification.map do |gem|
```

Comment:
I noticed that running `Gem::Specification.map{|x|x}` inside a simple `pry` terminal I get all gems installed in the system (including multiple versions of the same gem). This is not desirable, as this list is very long and uninformative to us.

I recommend we check if we are running inside a bundler environment before running `Gem::Specification.map`.
I wasn't able to find a definitive way to check if we are running inside a bundled environment.

Here's my "next best thing" suggestion:
```
begin
  !Bundler.bundle_path.nil?
rescue Bundler::GemfileNotFound
  false
end
```

This will return `true` when running inside a `bundle exec` or environment where bundler was loaded dynamically, and also with inline [bundler environments](https://bundler.io/guides/bundler_in_a_single_file_ruby_script.html).

Reply:
You should be able to get "loaded" versions with `Gem.loaded_specs`. That's what we need at a minimum. The following is what I get in our test environment:

```ruby
irb(main):013:0> pp Gem.loaded_specs.collect { |loaded_gem| [loaded_gem.first, loaded_gem.last.version.to_s] }.to_h
{"did_you_mean"=>"1.4.0",
 "bundler"=>"2.1.4",
 "uri"=>"0.10.0",
 "rake"=>"13.0.6",
 "addressable"=>"2.4.0",
 "thor"=>"1.2.1",
 "appraisal"=>"2.4.1",
 "ast"=>"2.4.2",
 "benchmark-ips"=>"2.10.0",
 "memory_profiler"=>"0.9.14",
 "benchmark-memory"=>"0.1.2",
 "debug_inspector"=>"1.1.0",
 "binding_of_caller"=>"0.8.0",
 "builder"=>"3.2.4",
 "byebug"=>"11.1.3",
 "climate_control"=>"0.2.0",
 "coderay"=>"1.1.3",
 "colorize"=>"0.8.1",
 "concurrent-ruby"=>"1.1.10",
 "rexml"=>"3.2.5",
 "crack"=>"0.4.5",
 "cri"=>"2.15.10",
 "debase-ruby_core_source"=>"0.10.16",
 "libddprof"=>"0.6.0.1.0",
 "ffi"=>"1.15.5",
 "libddwaf"=>"1.3.0.2.0",
 "msgpack"=>"1.5.2",
 "ddtrace"=>"1.1.0",
 "diff-lcs"=>"1.5.0",
 "docile"=>"1.4.0",
 "dogstatsd-ruby"=>"5.5.0",
 "google-protobuf"=>"3.19.1",
 "hashdiff"=>"1.0.1",
 "json-schema"=>"2.8.1",
 "method_source"=>"1.0.0",
 "opentracing"=>"0.5.0",
 "os"=>"1.1.4",
 "parallel"=>"1.20.1",
 "parser"=>"3.1.2.0",
 "pimpmychangelog"=>"0.1.3",
 "pry"=>"0.14.1",
 "pry-byebug"=>"3.8.0",
 "pry-stack_explorer"=>"0.4.12",
 "rainbow"=>"3.1.1",
 "rake-compiler"=>"1.2.0",
 "redcarpet"=>"3.5.1",
 "regexp_parser"=>"2.5.0",
 "rspec-support"=>"3.11.0",
 "rspec-core"=>"3.11.0",
 "rspec-expectations"=>"3.11.0",
 "rspec-mocks"=>"3.11.1",
 "rspec"=>"3.11.0",
 "rspec-collection_matchers"=>"1.2.0",
 "rspec_junit_formatter"=>"0.5.1",
 "rspec_n"=>"1.3.0",
 "rubocop-ast"=>"1.4.1",
 "ruby-progressbar"=>"1.11.0",
 "unicode-display_width"=>"2.1.0",
 "rubocop"=>"1.12.1",
 "rubocop-performance"=>"1.10.2",
 "rubocop-rspec"=>"2.2.0",
 "ruby-prof"=>"1.4.2",
 "simplecov-html"=>"0.12.3",
 "simplecov_json_formatter"=>"0.1.4",
 "simplecov"=>"0.21.2",
 "sorbet-static"=>"0.5.9672",
 "sorbet"=>"0.5.9672",
 "sorbet-runtime"=>"0.5.10090",
 "spoom"=>"1.1.11",
 "warning"=>"1.2.1",
 "webmock"=>"3.13.0",
 "webrick"=>"1.7.0",
 "yard"=>"0.9.28"}
```

Reply:
It may not be appropriate for the current version of this event, however, it might be nice if we could emit not just version loaded but version requirements as well (derived from the Gemfile from Bundler).

e.g. `version` would be `1.2.0` but `version_requirement` would be `~> 1.0`

Reply:
`Gem.loaded_specs` seems to work! 

At the moment, the only fields for the `Dependency` object are `name`, `version` and `hash` -- would it still be worthwhile to create a utility for capturing the version requirement / to save that for a future endeavor if/when the telemetry specs are updated to support this field?